docs/closure-compiler.md

```diff
@@ -0,0 +1,699 @@
+[easily-breakable client code](https://medium.com/@rufuszh90/effective-java-item-17-design-and-document-for-inheritance-or-else-prohibit-it-be6041719fbc) (note that this excerpt is taken from the book [Effective Java, 2nd Edition](https://www.pearsonhighered.com/program/Bloch-Effective-Java-2nd-Edition/PGM310651.html) by Joshua Bloch).
+
+The obvious exception to this rule is for classes that are intended to be subclassed. These should
+be well documented, and made `@abstract` if possible.
+
+#### Foundation classes must extend `MDCFoundation`
+
+Foundations must extend `MDCFoundation` parameterized by their respective adapter. The
+`defaultAdapter` must return an object with the correct adapter shape, type-cast to the type for
+the adapter.
+
+```js
+// foundation.js
+
+import MDCFoundation from '@material/base/foundation';
+import MDCComponentAdapter from './adapter';
+
+/** @final @extends {MDCFoundation<!MDCComponentAdapter>} */
+export default class MDCComponentFoundation extends MDCFoundation {
+  static get defaultAdapter() {
+    return /** @type {!MDCComponentAdapter} */ ({
+      addClass: () => {},
+      removeClass: () => {},
+    });
+  }
+}
+```
+
+#### Component classes must extend `MDCComponent`
+
+Components must extend `MDCComponent` parameterized by their respective foundation.
+
+```js
+// index.js
+
+import MDCComponent from '@material/base/component';
+import MDCComponentFoundation from './foundation';
+
+/** @final @extends {MDCComponent<!MDCComponentFoundation>} */
+export class MDCAwesomeComponent extends MDCComponent {
+  getDefaultFoundation() {
+    return new MDCComponentFoundation({
+      addClass: (className) => this.root_.classList.add(className),
+      removeClass: (className) => this.root_.classList.remove(className),
+    });
+  }
+}
+```
+
+#### @typedefs are always `let` declarations, always pascal case, and always end in `Type`
+
+```js
+// GOOD
+/**
+ * @typedef {{foo: string, bar: number}}
+ */
+let EventDataType;
+
+// GOOD
+/**
+ * @typedef {{foo: string, bar: number}}
+ */
+export let EventDataType;
+
+// BAD
+/**
+ * @typedef {{foo: string, bar: number}}
+ */
+MDCComponentFoundation.EventDataType;
+
+// BAD
+/**
+ * @typedef {{foo: string, bar: number}}
+ */
+let eventDataType;
+
+// BAD
+/**
+ * @typedef {{foo: string, bar: number}}
+ */
+let EventData;
+```
+
+Using this convention allows us to write tooling around handling these expressions, such as
+lint rule exceptions, and (in the future) code removal tools.
+
+### Objects that use event names or other external symbols as keys must be declared as `@dict` or `!Object<string, T>`.
+
+```js
+// GOOD
+/** @const {!Object<string, string>}  */
+const activationDeactivationPairs = {
+  'mousedown': 'mouseup',
+  'touchstart': 'touchend',
+};
+
+// GOOD
+/** @dict */
+window.settings = {
+  'windowObject': window,
+  'domReadyEvent': 'onready',
+  'scriptExecutionTimeoutMs': 3000,
+}
+
+// BAD
```

Comment:
It wasn't immediately obvious to me what was wrong with this. Then I realized that the keys weren't quoted, which is described below. Should the description come first? Or maybe the BAD comment should say why it's bad?